### PR TITLE
First pass at keras tests

### PR DESCRIPTION
**Pull request by vanpelt**
Lot's going on here.  Highlights include

1.  Improved developer documentation
2. Refactored the test directory to be more clear on utils
3. New notebook testing harness, still needs work but is running
4. Made compat.tempfile use the system library for all py3+
3. Added python 3.8 to CI / Tox
4. Added Keras Tests
5. Added Torch Tests
6. Added Scikit Tests

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -20,6 +28,15 @@
 DUMMY_API_KEY = '1824812581259009ca9981580f8f8a9012409eee'
 
 
+def debug(*args, **kwargs):
+    print("Open files during tests: ")
+    proc = psutil.Process()
+    print(proc.open_files())
```
Comment (vanpelt) [inline tests/conftest.py:34]: On mac we were getting too many open file errors.  Unfortunately this didn't serve to help as the open files are likely coming from not our process...

File: tests/conftest.py [hunk 2]
```diff
@@ -88,24 +105,70 @@ def mock_server():
...
+    #  server = Process(target=worker, args=(app, port))
+    #  server.start()
+    root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
+    path = os.path.join(root, "tests", "utils", "mock_server.py")
+    command = ["python", path]
+    env = os.environ
+    env["PORT"] = str(port)
+    env["PYTHONPATH"] = root
+    server = subprocess.Popen(command, env=env)
+    server.base_url = "http://localhost:%s" % port
```
Comment (vanpelt) [inline tests/conftest.py:122]: Windows couldn't run the process with multiprocessing, so I just launch a new process.  There's substantial overhead here, we should launch one server for all tests and make it maintain state by test.

File: tests/test_git_repo.py [hunk 3]
```diff
@@ -12,11 +12,6 @@
 from wandb.internal.git_repo import GitRepo
 
 
-#  TODO: make this work in CI for windows
-@pytest.mark.skipif(
-    platform.system() == "Windows",
-    reason="GitRepo lib doesn't work in circleci windows",
-)
 class TestGitRepo:
```
Comment (vanpelt) [inline tests/test_git_repo.py:15]: This *just worked which was suprising.

File: wandb/__init__.py [hunk 4]
```diff
@@ -87,6 +87,12 @@ def _is_internal_process():
 summary = _preinit.PreInitObject("wandb.summary")
 log = _preinit.PreInitCallable("wandb.log")
 join = _preinit.PreInitCallable("wandb.join")
+# record of patched libraries
+patched = {
+    "tensorboard": [],
+    "keras": [],
+    "gym": []
+}
```
Comment (vanpelt) [inline wandb/__init__.py:95]: This fixes the keras callback for TF > 2

File: wandb/__init__.py [hunk 5]
```diff
@@ -98,6 +104,11 @@ def ensure_configured():
     global api
     api = InternalApi()
 
+def set_trace():
+    import pdb  # TODO: support other debuggers
+    #  frame = sys._getframe().f_back
+    pdb.set_trace() # TODO: pass the parent stack...
```
Comment (vanpelt) [inline wandb/__init__.py:110]: I should probably rip this out, but I think there are some cool things we could do here.  pytest exposes `pytest.set_trace` that stops mirroring io and adds some nice stuff.  Someday we'll do some cool stuff here.

File: wandb/compat/tempfile.py [hunk 6]
```diff
@@ -10,74 +10,83 @@
 import warnings as _warnings
 from shutil import rmtree as _rmtree
 
-from .weakref import finalize
-
```
Comment (vanpelt) [inline wandb/compat/tempfile.py:14]: I had thought too many open file woes were caused by this sketchy implementation so I tried to ensure we used the system TemporaryDirectory for anything but py2.

File: wandb/lib/server.py [hunk 7]
```diff
@@ -16,8 +16,8 @@ class Server(object):
     def __init__(self, api=None):
         self._api = api or InternalApi()
         self._error_network = None
-        self._viewer = None
-        self._flags = None
+        self._viewer = {}
+        self._flags = {}
```
Comment (vanpelt) [inline wandb/lib/server.py:20]: There are places in the code that assume these are set, I think this is the right thing to do.

File: wandb/__init__.py (hunk 5 above)
Comment (adrnswanberg) [reply wandb/__init__.py:110]: this is so that users can call `wandb.set_trace` in their code while debugging?

**Pull request merged**